recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+                with torch.cuda.amp.autocast():
+                    outputs = self.compute_forward(batch, Stage.TRAIN)
+                    objectives = self.compute_objectives(
+                        outputs, batch, Stage.TRAIN
+                    )
+                loss = objectives["loss"]
+                if self.hparams.diversity_loss_weight == 0.0:
+                    backprop_loss = loss
+                else:
+                    backprop_loss = (
+                        loss
+                        + objectives["diversity_loss"] * self.hparams.diversity_loss_weight# * objectives["num_masked"]
+                        + objectives["latent_l2"] * self.hparams.latent_l2_penalty_weight
+                    )
+                self.scaler.scale(
+                    backprop_loss / self.grad_accumulation_factor
+                ).backward()
+
+                objectives["total_loss"] = backprop_loss.detach()
+                if should_step:
+                    self.scaler.unscale_(self.optimizer)
+                    if self.check_gradients(loss):
+                        self.scaler.step(self.optimizer)
+                    self.scaler.update()
+                    self.optimizer_step += 1
+                    self.optimizer.zero_grad()
+        else:
+
+            with self.modules.model.no_sync() if not should_step else ExitStack():
+                outputs = self.compute_forward(batch, Stage.TRAIN)
+                objectives = self.compute_objectives(
+                    outputs, batch, Stage.TRAIN
+                )
+                loss = objectives["loss"]
+                if self.hparams.diversity_loss_weight == 0.0:
+                    backprop_loss = loss
+                else:
+                    backprop_loss = (
+                        loss
+                        + objectives["diversity_loss"] * self.hparams.diversity_loss_weight
+                        + objectives["latent_l2"] * self.hparams.latent_l2_penalty_weight
+                    )
+                (backprop_loss / self.grad_accumulation_factor).backward()
+                objectives["total_loss"] = backprop_loss.detach()
+                if should_step:
+                    self.optimizer.step()
+                    self.optimizer.zero_grad()
+                    self.optimizer_step += 1
+
+        if should_step:
+            if isinstance(
+                self.modules.model, DistributedDataParallel
+            ):
+                w2v_model = self.modules.model.module
+            else:
+                w2v_model = self.modules.model
+            if w2v_model.quantiser is not None:
+                w2v_model.quantiser.update_temp(
+                    self.optimizer_step
+                )
+
+            if (
+                hasattr(self.hparams, "log_interval")
+                and self.optimizer_step % self.hparams.log_interval
+                == 0
+            ):
+
+                log_dct = {
```

Comment:
I guess that this is mandatory for your logging, but it's very hard to follow. Is it mandatory due to wandb?

Reply:
Do you mean it's unclear because it's not clear why I'm doing this? The idea here is I just log everything (it's not needed because of wandb). Another option is to be more explicit like `logger.info(f'Step: {self.optimizer_step} - loss: {objectives["loss"]} - prob_perplexity: {objectives["prob_perplexity"]} - latent_l2: {objectives["latent_l2"} - lr: {lr} - temp: {temp} ....)`

But that is a lot more verbose and less robust if I want to add or remove something from logging.